### PR TITLE
USER-STORY-3217: Add instrumentation and audit to personal information service and letters service

### PR DESCRIPTION
**Pull request by carolinegeorge-gov**
### Description

* Added instrumentation and audit to personal-information and letters service

### Related Azure Boards Work Items

[AB#3217](https://dev.azure.com/DTS-STN/1fc40a8f-28cf-47bc-b6e4-1c234bd06177/_workitems/edit/3217)
### Checklist

- [X] I have tested the changes locally
- [X] I have updated the documentation if necessary
- [X] I have checked that my code follows the project's coding style by running `npm run format:check`
- [X] I have checked that my code contains no linting errors by running `npm run lint`
- [X] I have checked that my code contains no type errors by running `npm run typecheck`
- [X] I have checked that all unit tests pass by running `npm run test:unit -- run`
- [X] I have checked that all e2e tests pass by running `npm run test:e2e`

### CODE REVIEW

File: frontend/app/services/letters-service.server.ts [hunk 1]
```diff
@@ -105,6 +118,7 @@ function createLettersService() {
     });
 
     if (!response.ok) {
+      instrumentationService.countHttpStatus('http.client.interop-api.letters.posts', response.status);
```
Comment (gregory-j-baker) [inline frontend/app/services/letters-service.server.ts:121]: I you move this line of code outside of the `if (..)`, then you would only need to call `.countHttpStatus(..)` once in this function.

This comment applies to other similar code in this PR.

File: frontend/app/services/personal-information-service.server.ts [hunk 2]
```diff
@@ -1,6 +1,8 @@
 import moize from 'moize';
 import { z } from 'zod';
 
+import { getAuditService } from './audit-service.server';
+import { getInstrumentationService } from './instrumentation-service.server';
```
Comment (gregory-j-baker) [inline frontend/app/services/personal-information-service.server.ts:5]:
```suggestion
import { getAuditService } from '~/services/audit-service.server';
import { getInstrumentationService } from '~/services/instrumentation-service.server';
```

File: frontend/app/services/personal-information-service.server.ts [hunk 3]
```diff
@@ -212,6 +214,9 @@ function createPersonalInformationService() {
   async function getPersonalInformation(personalSinId: string) {
     const curentPersonalInformation = createClientInfo(personalSinId);
     const url = `${INTEROP_POWERPLATFORM_API_BASE_URI}/dental-care/applicant-information/dts/v1/applicant/`;
+    const auditService = getAuditService();
+    const instrumentationService = getInstrumentationService();
+    auditService.audit('personal-information.get', { userId: 'anonymous' });
```
Comment (gregory-j-baker) [inline frontend/app/services/personal-information-service.server.ts:219]: We should log either the `clientId` here, or some other identifier that we can use to find the audit event in the future.

**Review by gregory-j-baker — CHANGES_REQUESTED** [review]

File: frontend/app/services/letters-service.server.ts [hunk 4]
```diff
@@ -45,6 +47,10 @@ function createLettersService() {
    * @returns array of letters given the clientId with optional sort parameter
    */
   async function getLetters(clientId: string, sortOrder: 'asc' | 'desc' = 'desc') {
+    const auditService = getAuditService();
+    const instrumentationService = getInstrumentationService();
+
+    auditService.audit('letters.get', { userId: 'anonymous' });
```
[the remark this answers is not in the document]
Comment (carolinegeorge-gov) [reply frontend/app/services/letters-service.server.ts]: That's right. Mixed it up with another implementation. Fixed: fe83f7777d7d1765885535ec99e2b31d758b8fb6

File: frontend/app/services/letters-service.server.ts (hunk 1 above)
Comment (carolinegeorge-gov) [reply frontend/app/services/letters-service.server.ts:123]: Would I just use the one `instrumentationService.countHttpStatus('http.client.interop-api.letters.posts', response.status)` then? (and not the `instrumentationService.countHttpStatus('http.client.interop-api.letters.posts', 200);` outside it)

File: frontend/app/services/personal-information-service.server.ts (hunk 3 above)
Comment (carolinegeorge-gov) [reply frontend/app/services/personal-information-service.server.ts]: Fixed: fe83f7777d7d1765885535ec99e2b31d758b8fb6

File: frontend/app/services/personal-information-service.server.ts (hunk 2 above)
Comment (carolinegeorge-gov) [reply frontend/app/services/personal-information-service.server.ts]: Fixed: fe83f7777d7d1765885535ec99e2b31d758b8fb6

File: frontend/app/services/letters-service.server.ts [hunk 5]
```diff
@@ -95,6 +104,10 @@ function createLettersService() {
   async function getPdf(letterId: string) {
     const url = new URL(`${INTEROP_CCT_API_BASE_URI}/dental-care/client-letters/cct/v1/GetPdfByLetterId`);
     url.searchParams.set('id', letterId);
+    const auditService = getAuditService();
+    const instrumentationService = getInstrumentationService();
+
+    auditService.audit('letters.get', { userId: 'anonymous' });
```
[the remark this answers is not in the document]
Comment (carolinegeorge-gov) [reply frontend/app/services/letters-service.server.ts]: Fixed: dc613150334e152a1a5f2812b2d27980494e3972

File: frontend/app/services/letters-service.server.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (nicholas-ly) [reply frontend/app/services/letters-service.server.ts]: We seem to be using the `idToken.sub` for auditing in the authenticated space, which is different from what is passed into this function (the `clientNumber`). I wonder if we should be consistent and stick with `idToken.sub` instead.

[the remark this answers is not in the document]
Comment (gregory-j-baker) [reply frontend/app/services/letters-service.server.ts]: I agree with you @nicholas-ly. I believe we were asked to use the RAOIDC `sub` claim as the user id.

File: frontend/app/services/letters-service.server.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (gregory-j-baker) [reply frontend/app/services/letters-service.server.ts]: @nicholas-ly same comment as above for the `sub` claim?

File: frontend/app/services/letters-service.server.ts (hunk 1 above)
Comment (gregory-j-baker) [reply frontend/app/services/letters-service.server.ts:123]: Yes.. that is correct.

File: frontend/app/services/personal-information-service.server.ts (hunk 3 above)
Comment (gregory-j-baker) [reply frontend/app/services/personal-information-service.server.ts]: @nicholas-ly again.. do you think this should be the RAOIDC `sub` claim?

File: frontend/app/services/letters-service.server.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (nicholas-ly) [reply frontend/app/services/letters-service.server.ts]: Yes, let's audit with the `sub` claim as the `userId`.

File: frontend/app/services/letters-service.server.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (carolinegeorge-gov) [reply frontend/app/services/letters-service.server.ts]: Fixed: b54fdc1071cc2f7076d56ba903332d1f0b525eed

File: frontend/app/services/letters-service.server.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (carolinegeorge-gov) [reply frontend/app/services/letters-service.server.ts]: Fixed: b54fdc1071cc2f7076d56ba903332d1f0b525eed

File: frontend/app/services/letters-service.server.ts (hunk 1 above)
Comment (carolinegeorge-gov) [reply frontend/app/services/letters-service.server.ts]: Fixed: b54fdc1071cc2f7076d56ba903332d1f0b525eed

File: frontend/app/services/personal-information-service.server.ts (hunk 3 above)
Comment (carolinegeorge-gov) [reply frontend/app/services/personal-information-service.server.ts]: Fixed as per previous discussions: b54fdc1071cc2f7076d56ba903332d1f0b525eed

File: frontend/app/services/letters-service.server.ts [hunk 6]
```diff
@@ -46,7 +48,11 @@ function createLettersService() {
   /**
    * @returns array of letters given the clientId with optional sort parameter
    */
-  async function getLetters(clientId: string, sortOrder: 'asc' | 'desc' = 'desc') {
+  async function getLetters(clientId: string, userId: string, sortOrder: 'asc' | 'desc' = 'desc') {
+    const auditService = getAuditService();
+    const instrumentationService = getInstrumentationService();
+    auditService.audit('letters.get', { userId: userId });
```
Comment (gregory-j-baker) [inline frontend/app/services/letters-service.server.ts:54]: **Nitpick:** (sorry)

```suggestion
    auditService.audit('letters.get', { userId });
```

File: frontend/app/services/letters-service.server.ts [hunk 7]
```diff
@@ -94,9 +101,13 @@ function createLettersService() {
...
    * @returns a promise that resolves to a base64 encoded string representing the PDF document
    */
-  async function getPdf(letterId: string) {
+  async function getPdf(letterId: string, userId: string) {
     const url = new URL(`${INTEROP_CCT_API_BASE_URI ?? INTEROP_API_BASE_URI}/dental-care/client-letters/cct/v1/GetPdfByLetterId`);
     url.searchParams.set('id', letterId);
+    const auditService = getAuditService();
+    const instrumentationService = getInstrumentationService();
+
+    auditService.audit('letters.get', { letterId: letterId, userId: userId });
```
Comment (gregory-j-baker) [inline frontend/app/services/letters-service.server.ts:110]:
```suggestion
    auditService.audit('letters.get', { letterId, userId });
```

File: frontend/app/services/personal-information-service.server.ts [hunk 8]
```diff
@@ -215,9 +217,12 @@ function createPersonalInformationService() {
     return { Applicant: { PersonSINIdentification: { IdentificationID: personalSinId } } };
   }
 
-  async function getPersonalInformation(personalSinId: string) {
+  async function getPersonalInformation(personalSinId: string, userId: string) {
     const curentPersonalInformation = createClientInfo(personalSinId);
     const url = `${INTEROP_APPLICANT_API_BASE_URI ?? INTEROP_API_BASE_URI}/dental-care/applicant-information/dts/v1/applicant/`;
+    const auditService = getAuditService();
+    const instrumentationService = getInstrumentationService();
+    auditService.audit('personal-information.get', { userId: userId });
```
Comment (gregory-j-baker) [inline frontend/app/services/personal-information-service.server.ts:225]:
```suggestion
    auditService.audit('personal-information.get', { userId });
```

**Review by gregory-j-baker — DISMISSED** [review]

File: frontend/app/services/letters-service.server.ts (hunk 7 above)
Comment (carolinegeorge-gov) [reply frontend/app/services/letters-service.server.ts]: Fixed: 3f1790498d19e939c0427f00fb02e815393caa47

File: frontend/app/services/personal-information-service.server.ts (hunk 8 above)
Comment (carolinegeorge-gov) [reply frontend/app/services/personal-information-service.server.ts]: Fixed: 3f1790498d19e939c0427f00fb02e815393caa47

File: frontend/app/routes/$lang+/_protected+/letters+/$id.download.tsx [hunk 9]
```diff
@@ -38,8 +40,11 @@ export async function loader({ context: { session }, params, request }: LoaderFu
   }
   const locale = getLocale(request);
   const documentName = sanitize(getNameByLanguage(locale, letterType) ?? '');
+  const userInfoToken: UserinfoToken = session.get('userInfoToken');
+  const personalInformationRouteHelpers = getPersonalInformationRouteHelpers();
+  const personalInformation = await personalInformationRouteHelpers.getPersonalInformation(userInfoToken, params, request, session);
 
-  const pdfBytes = await lettersService.getPdf(params.id);
+  const pdfBytes = personalInformation.clientNumber ? await lettersService.getPdf(params.id, personalInformation.clientNumber) : await lettersService.getPdf(params.id, 'no client id');
```
[the remark this answers is not in the document]
Comment (carolinegeorge-gov) [reply frontend/app/routes/$lang+/_protected+/letters+/$id.download.tsx]: Fixed: 3f1790498d19e939c0427f00fb02e815393caa47

File: frontend/app/services/letters-service.server.ts [hunk 10]
```diff
@@ -58,6 +64,7 @@ function createLettersService() {
       },
     });
 
+    instrumentationService.countHttpStatus('http.client.interop-api.letters.posts', response.status);
```
[the remark this answers is not in the document]
Comment (carolinegeorge-gov) [reply frontend/app/services/letters-service.server.ts:67]: Fixed: 3f1790498d19e939c0427f00fb02e815393caa47

File: frontend/app/services/letters-service.server.ts [hunk 11]
```diff
@@ -106,6 +117,7 @@ function createLettersService() {
...
```
[the remark this answers is not in the document]
Comment (carolinegeorge-gov) [reply frontend/app/services/letters-service.server.ts:120]: Fixed: 3f1790498d19e939c0427f00fb02e815393caa47

File: frontend/app/services/letters-service.server.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (nicholas-ly) [reply frontend/app/services/letters-service.server.ts:67]: I still see `instrumentationService.countHttpStatus('http.client.interop-api.letters.posts', response.status);` (same with the other comments dealing with `instrumentationService`). Maybe the change was not pushed?

File: frontend/app/services/personal-information-service.server.ts [hunk 12]
```diff
@@ -229,6 +234,8 @@ function createPersonalInformationService() {
       },
     });
 
+    instrumentationService.countHttpStatus('http.client.interop-api.personal-information.posts', response.status);
```
[the remark this answers is not in the document]
Comment (carolinegeorge-gov) [reply frontend/app/services/personal-information-service.server.ts]: Fixed: 0a7c3610c748ed0c0265e8ecdf3b9a0916069e44

File: frontend/app/services/letters-service.server.ts [hunk 13]
```diff
@@ -46,7 +48,11 @@ function createLettersService() {
   /**
    * @returns array of letters given the clientId with optional sort parameter
    */
-  async function getLetters(clientId: string, sortOrder: 'asc' | 'desc' = 'desc') {
+  async function getLetters(clientId: string, userId: string, sortOrder: 'asc' | 'desc' = 'desc') {
```
[the remark this answers is not in the document]
Comment (carolinegeorge-gov) [reply frontend/app/services/letters-service.server.ts:51]: Fixed: 0a7c3610c748ed0c0265e8ecdf3b9a0916069e44

File: frontend/app/services/letters-service.server.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (carolinegeorge-gov) [reply frontend/app/services/letters-service.server.ts]: I had used the wrong commit id. Is it there in 0a7c3610c748ed0c0265e8ecdf3b9a0916069e44?

[the remark this answers is not in the document]
Comment (nicholas-ly) [reply frontend/app/services/letters-service.server.ts]: Yep! However the attribute name is currently `http.client.interop-api.get-pdf-by-letter-id.gets` for the `/GetDocInfoByClientId` endpoint. For consistency, the name should be `http.client.interop-api.get-doc-info-by-client-id.gets`.

File: frontend/app/services/letters-service.server.ts [hunk 14]
```diff
@@ -94,9 +101,13 @@ function createLettersService() {
...
    * @returns a promise that resolves to a base64 encoded string representing the PDF document
    */
-  async function getPdf(letterId: string) {
+  async function getPdf(letterId: string, userId: string) {
     const url = new URL(`${INTEROP_CCT_API_BASE_URI ?? INTEROP_API_BASE_URI}/dental-care/client-letters/cct/v1/GetPdfByLetterId`);
     url.searchParams.set('id', letterId);
+    const auditService = getAuditService();
+    const instrumentationService = getInstrumentationService();
+
+    auditService.audit('letters.get', { letterId, userId });
```
Comment (nicholas-ly) [inline frontend/app/services/letters-service.server.ts:110]: Sorry, I missed this one initially. We should distinguish calls between `getPdf(..)` and `getLetters(..)`

```suggestion
    auditService.audit('pdf.get', { letterId, userId });
```

File: frontend/app/services/letters-service.server.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (carolinegeorge-gov) [reply frontend/app/services/letters-service.server.ts]: Fixed, Is it for both places?: 237eb282fc13194799506aa1295a4b30bd775b8f

File: frontend/app/services/letters-service.server.ts (hunk 14 above)
Comment (carolinegeorge-gov) [reply frontend/app/services/letters-service.server.ts]: Fixed: 237eb282fc13194799506aa1295a4b30bd775b8f

File: frontend/app/services/letters-service.server.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (nicholas-ly) [reply frontend/app/services/letters-service.server.ts]: The name would not apply for both API calls. 

One API endpoint is `GET /GetPdfByLetterId`, which would use `http.client.interop-api.get-pdf-by-letter-id.gets`

The other is `GET /GetDocInfoByClientId`, which would use `http.client.interop-api.get-doc-info-by-client-id.gets`

[the remark this answers is not in the document]
Comment (carolinegeorge-gov) [reply frontend/app/services/letters-service.server.ts]: Got it. Just wanted to be sure what you meant by consistency. Fixed 6fcdf46041a4567a7450afdbad81e98a137c1936

**Review by nicholas-ly — APPROVED** [review]

**Pull request merged**